tensorflow/python/ops/script_ops.py

```diff
@@ -129,7 +129,108 @@ def __del__(self):
+  TensorFlow graph that invokes the `np.sinh()` NumPy function as a operation
+  in the graph:
+
+  ```python
+  def my_func(x):
+    # x will be a numpy array with the contents of the placeholder below
+    return np.sinh(x)
+  inp = tf.placeholder(tf.float32)
+  y = tf.py_func(my_func, [inp], tf.float32)
+  ```
+
+
+  **N.B.** The `tf.py_func()` operation has the following known limitations:
+
+  * The body of the function (i.e. `func`) will not be serialized in a
+    `GraphDef`. Therefore, you should not use this function if you need to
+    serialize your model and restore it in a different environment.
+
+  * The operation must run in the same address space as the Python program
+    that calls `tf.py_func()`. If you are using distributed TensorFlow, you
+    must run a `tf.train.Server` in the same process as the program that calls
+    `tf.py_func()` and you must pin the created operation to a device in that
+    server (e.g. using `with tf.device():`).
+
+  Args:
+    func: A Python function, which accepts a list of NumPy `ndarray` objects
+      having element types that match the corresponding `tf.Tensor` objects
+      in `inp`, and returns a list of `ndarray` objects (or a single `ndarray`)
+      having element types that match the corresponding values in `Tout`.
+    args: A list of `Tensor` objects.
+    kwargs: A dict with `Tensor` objects as values.
+    output_types: A nested structure of tensorflow data types or a single
+      tensorflow data type if there is only one, indicating what `func` returns.
+    output_shapes: Same as output_types, except the types are replaces with
+      shapes (optional).
+    stateful: (Boolean.) If True, the function should be considered stateful.
+      If a function is stateless, when given the same input it will return the
+      same output and have no observable side effects. Optimizations such as
+      common subexpression elimination are only performed on stateless
+      operations.
+    name: A name for the operation (optional).
+
+
+  """
+  from tensorflow.python.framework import tensor_shape
+  from tensorflow.python.util import nest
+  # Note: tf.data.Dataset.from_generator use tensorflow.python.data.util.nest
+
+  if not isinstance(args, (list, tuple)):
+    raise TypeError('args must be list and not {}. args: {}'.format(
+      type(args), args))
+
+  if not isinstance(kwargs, dict):
+    raise TypeError('kwargs must be dict and not {}. args: {}'.format(
+      type(kwargs), kwargs))
+
+  # For dynamic type inference use callable output_types and output_shapes
+  if callable(output_types):
+    # If callable, assume same signature and call with tensors and get the types
+    output_types = output_types(*args, **kwargs)
+  if callable(output_shapes):
+      # If callable, assume same signature and call with tensors and get the shapes
+      output_shapes = output_shapes(*args, **kwargs)
+
+  flat_output_types = nest.flatten(output_types)
```

Comment:
So my understanding is that the main change here is you call nest.flatten and nest.pack_sequence_as around the user code in a py_func. Everything else is debugging / diagnostics, right?

If so, I think this can live in utility code instead of in the tensorflow library proper. Maybe you want to add a contrib.py_func with this API or add this py func wrapper in a third-party library somewhere?

Reply:
You are right. The main changed are allow kwargs as input and arbitrary returns with shapes (earlier only list/scalar without shape).
Another (not so important) point is to allow shape/type inference from the input.(Useful for decorators)
If you prefer `contrib.py_func` I can move the code to contrib.
If you prefer a third-party library, I will keep the code in a private repo.
 
When you need a use case inside of tf for this modification, look in `tf.data.Dataset.from_generator`.
The code there was the starting point.

Reply:
I agree the dataset API for py_func is friendlier to use and I like exposing it. We could consider either tf.py_func_v2, or contrib.

@martinwicke for API review as to whether we prefer _v2 or contrib